ergotree-ir/src/types/stype_companion.rs

```diff
@@ -22,66 +23,103 @@ pub(crate) struct STypeCompanionHead {
-        } else if type_id == scoll::S_COLL_TYPE_COMPANION.type_id() {
-            &scoll::S_COLL_TYPE_COMPANION
-        } else if type_id == sgroup_elem::S_GROUP_ELEM_TYPE_COMPANION.type_id() {
-            &sgroup_elem::S_GROUP_ELEM_TYPE_COMPANION
-        } else if type_id == sglobal::S_GLOBAL_TYPE_COMPANION.type_id() {
-            &sglobal::S_GLOBAL_TYPE_COMPANION
-        } else if type_id == sheader::S_HEADER_TYPE_COMPANION.type_id() {
-            &sheader::S_HEADER_TYPE_COMPANION
-        } else if type_id == spreheader::S_PRE_HEADER_TYPE_COMPANION.type_id() {
-            &spreheader::S_PRE_HEADER_TYPE_COMPANION
-        } else if type_id == soption::S_OPTION_TYPE_COMPANION.type_id() {
-            &soption::S_OPTION_TYPE_COMPANION
-        } else {
-            return Err(SigmaParsingError::NotImplementedYet(format!(
-                "cannot find STypeCompanion for {0:?} type id",
-                type_id,
-            )));
-        })
+    fn method_desc<'a>(&'a self) -> &'a Vec<&'static SMethodDesc> {
+        match self {
+            STypeCompanion::Context => &*scontext::METHOD_DESC,
+            STypeCompanion::Box => &*sbox::METHOD_DESC,
+            STypeCompanion::Coll => &*scoll::METHOD_DESC,
+            STypeCompanion::GroupElem => &*sgroup_elem::METHOD_DESC,
+            STypeCompanion::Global => &*sglobal::METHOD_DESC,
+            STypeCompanion::Header => &*sheader::METHOD_DESC,
+            STypeCompanion::PreHeader => &*spreheader::METHOD_DESC,
+            STypeCompanion::Option => &*soption::METHOD_DESC,
+        }
     }
 
     /// Get method signature for this object by a method id
-    pub fn method_by_id(&'static self, method_id: &MethodId) -> Option<SMethod> {
-        self.methods
+    pub fn method_by_id(&self, method_id: &MethodId) -> Option<SMethod> {
+        self.method_desc()
             .iter()
             .find(|m| m.method_id == *method_id)
-            .map(|m| m.as_method(self))
+            .map(|m| m.as_method(*self))
     }
 
     /// Get list of method signatures for this object's type companion
-    pub fn methods(&'static self) -> Vec<SMethod> {
-        self.methods.iter().map(|m| m.as_method(self)).collect()
+    pub fn methods(&self) -> Vec<SMethod> {
+        self.method_desc()
+            .iter()
+            .map(|m| m.as_method(*self))
+            .collect()
     }
 
     /// Get object type id for this type companion
-    pub fn type_id(&'static self) -> TypeCode {
-        self.head.type_id
+    pub fn type_id(&self) -> TypeCode {
+        self.s_type_companion_head().type_id
     }
 
     /// Get object's type name
-    pub fn type_name(&'static self) -> &'static str {
-        self.head.type_name
+    pub fn type_name(&self) -> &'static str {
+        self.s_type_companion_head().type_name
+    }
+}
+
+impl TryFrom<TypeCode> for STypeCompanion {
+    type Error = SigmaParsingError;
+    fn try_from(value: TypeCode) -> Result<Self, Self::Error> {
```

Comment:
@greenhat this code manually covers all the enum variants. We can get more compiler assurance by using the `strum` crate (https://crates.io/crates/strum), which has the ability to iterate over all enum variants. What do you think?

Reply:
`strum` is neat and I agree to pull it in. I'm wondering if we could ditch `STypeCompanionHead` completely. What do you think of adding `STypeCompanion::type_code()` to return type code and `STypeCompanion::type_name()` to return type name?

Reply:
Sure, I'll make the changes.